pkg/kubernetes/version.go

```diff
@@ -0,0 +1,100 @@
+package kubernetes
+
+import (
+	"fmt"
+	"sort"
+	"strconv"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+const (
+	// https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/
+	AnnotationSpinnakerArtifactVersion = `artifact.spinnaker.io/version`
+	AnnotationSpinnakerMonikerSequence = `moniker.spinnaker.io/sequence`
+)
+
+type SpinnakerVersion struct {
+	Long  string
+	Short string
+}
+
+//Create a similar function for labels
+func (c *controller) AddSpinnakerVersionAnnotations(u *unstructured.Unstructured, application string, version SpinnakerVersion) error {
+	// var err error
+
+	annotate(u, AnnotationSpinnakerArtifactVersion, version.Long)
+	annotate(u, AnnotationSpinnakerMonikerSequence, version.Short)
+
+	//ToDo
+	// Add spinnaker versioning labels and annotations.
+	// 		.metadata.annotations:
+	//   artifact.spinnaker.io/version: vNNN
+	//   moniker.spinnaker.io/sequence: "N"
+	// .metadata.labels:
+	//   moniker.spinnaker.io/sequence: "N"
+	// .spec.template.metadata.annotations:
+	//   artifact.spinnaker.io/version: vNNN
+	//   moniker.spinnaker.io/sequence: "N"
+	// .spec.template.metadata.labels:
+	//   moniker.spinnaker.io/sequence: "N"
+
+	return nil
+}
+
+func (c *controller) GetCurrentVersion(ul *unstructured.UnstructuredList, kind, name string) string {
+
+	cluster := kind + " " + name
+	currentVersion := "0"
+	if len(ul.Items) == 0 {
+		return currentVersion
+	}
+	// Filter out all unassociated objects based on the moniker.spinnaker.io/cluster annotation.
+	results := FilterOnCluster(ul.Items, cluster)
+	if len(results) == 0 {
+		return currentVersion
+	}
+
+	//filter out empty moniker.spinnaker.io/sequence labels
+	results = FilterWhereLabelDoesNotExist(results, LabelSpinnakerSequence)
+	if len(results) == 0 {
+		return currentVersion
+	}
+
+	// For now, we sort on creation timestamp to grab the manifest.
+	sort.Slice(results, func(i, j int) bool {
+		return results[i].GetCreationTimestamp().String() > results[j].GetCreationTimestamp().String()
+	})
+	currentVersion = results[0].GetResourceVersion()
+
+	return currentVersion
+}
+
+func (c *controller) IsVersioned(u *unstructured.Unstructured) bool {
+	annotations := u.GetAnnotations()
+	if annotations != nil {
+		if _, ok := annotations[AnnotationSpinnakerStrategyVersioned]; ok {
```

Comment:
This shouldn't return true just because the annotation exists - it also needs to check if the annotation is set to `true`.